dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerDtoValidationHook.java

```diff
@@ -0,0 +1,323 @@
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+import static org.hisp.dhis.tracker.validation.hooks.TrackerImporterAssertErrors.ATTRIBUTE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.TrackerImporterAssertErrors.DATE_STRING_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.TrackerImporterAssertErrors.GEOMETRY_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.TrackerImporterAssertErrors.TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+public abstract class AbstractTrackerDtoValidationHook
+    implements TrackerValidationHook
+{
+    private int order = Ordered.LOWEST_PRECEDENCE;
+
+    @Override
+    public int getOrder()
+    {
+        return order;
+    }
+
+    public void setOrder( int order )
+    {
+        this.order = order;
+    }
+
+    public static final String IMPLEMENTING_CLASS_FAIL_TO_OVERRIDE_THIS_METHOD = "Implementing class fail to override this method!";
+
+    @Autowired
+    protected TrackedEntityAttributeService teAttrService;
+
+    @Autowired
+    private TrackedEntityCommentService commentService;
+
+    private final TrackerImportStrategy strategy;
+
+    private final boolean removeOnError;
+
+    private final Class<?> dtoTypeClass;
+
+    public AbstractTrackerDtoValidationHook()
+    {
+        this.removeOnError = true;
+        this.dtoTypeClass = null;
+        this.strategy = null;
+    }
+
+    public <T extends TrackerDto> AbstractTrackerDtoValidationHook( Class<T> dtoClass, TrackerImportStrategy strategy )
+    {
+        this.removeOnError = false;
+        this.dtoTypeClass = dtoClass;
+        this.strategy = strategy;
+    }
+
+    public void validateEvent( ValidationErrorReporter reporter, Event event )
+    {
+        throw new IllegalStateException( IMPLEMENTING_CLASS_FAIL_TO_OVERRIDE_THIS_METHOD );
+    }
+
+    public void validateEnrollment( ValidationErrorReporter reporter, Enrollment enrollment )
+    {
+        throw new IllegalStateException( IMPLEMENTING_CLASS_FAIL_TO_OVERRIDE_THIS_METHOD );
+    }
+
+    public void validateTrackedEntity( ValidationErrorReporter reporter, TrackedEntity tei )
+    {
+        throw new IllegalStateException( IMPLEMENTING_CLASS_FAIL_TO_OVERRIDE_THIS_METHOD );
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerImportValidationContext context )
+    {
+        TrackerBundle bundle = context.getBundle();
+
+        ValidationErrorReporter reporter = new ValidationErrorReporter( context, this.getClass() );
+
+        if ( this.strategy != null )
+        {
+            TrackerImportStrategy importStrategy = bundle.getImportStrategy();
+
+            if ( importStrategy.isDelete() && !this.strategy.isDelete() )
+            {
+                return reporter.getReportList();
+            }
+        }
+
+        if ( dtoTypeClass == null || dtoTypeClass.equals( TrackedEntity.class ) )
```

Comment:
I think we can be optimized by using `Map<Class, ValidationFunction<T extends TrackerDto>`.

Reply:
Hi, I don't like this method and am very happy to get some points on how to improve it. I am not quite sure what you mean with the map optimization you mention here. Can you please explain at bit more?

Reply:
I have improved a little on it now. I am not sure it was the same as you had in mind :)

Reply:
yes, that's way better now :)